### PR TITLE
feat(react): render variant hiding contexts from users

### PR DESCRIPTION
**Pull request by DJCordhose**
**Proposal for a simplified alternative to the existing react based rendering.**

Objective: hide context (a truly advanced feature) from the react developer

Instead of passing a context to render, you can now pass a configuration object that also describes which additional rendering features you need, like:

 ```javascript
import { reduxExtension } from 'hops-redux';

export default render(<App />, {
  extensions: [reduxExtension({ reducers })],
});
```

Each react based package now exports itself as an extension.

**This is a non-breaking change, if you want to you can still pass a context.**

### CODE REVIEW

File: packages/react/dom.js [hunk 1]
```diff
@@ -67,4 +69,18 @@ exports.render = function(reactElement, context) {
   };
 };
 
+exports.render = function(reactElement, contextOrConfig) {
+  var context;
+  if (typeof contextOrConfig === 'function') {
```
Comment (dmbch) [inline packages/react/dom.js]: I am afraid we are going to need a different check here - `contextOrConfig` could be a context instance here, but I would not really expect a function/constructor.

File: packages/react/node.js [hunk 2]
```diff
@@ -109,4 +111,18 @@ exports.render = function(reactElement, _context) {
   };
 };
 
+exports.render = function(reactElement, contextOrConfig) {
+  var context;
+  if (typeof contextOrConfig === 'function') {
```
Comment (dmbch) [inline packages/react/node.js]: Same here - `contextOrConfig` should not really be a function...

File: packages/react/lib/common.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+) {
+  var mergedConfig = Object.assign({}, config);
+  delete mergedConfig.extensions;
+  var mergedContexts = [ReactContext];
+  if (config.extensions) {
+    config.extensions.forEach(function(extension) {
+      var context = extension.context;
+      mergedContexts.push(context);
+      var extensionConfig = extension.config;
+      mergedConfig = Object.assign(mergedConfig, extensionConfig);
```
Comment (dmbch) [inline packages/react/lib/common.js]: I think I would rather `bind` the extension specific options to their respective constructor, something like..:

```
var config = Object.assign({}, mergedConfig, extensionConfig);
mergedContexts.push(
  Function.prototype.bind.call(context, context, config)
);
```

Comment (dmbch) [reply packages/react/lib/common.js]: ...but that is clearly suboptimal, as well (cloning...)

File: packages/react/dom.js (hunk 1 above)
Comment (dmbch) [reply packages/react/dom.js]: I think we ought to use the brand new [`isMixinable()`](https://www.npmjs.com/package/mixinable#defineismixinableobject) function here - and we could additionally check whether we have a proper `Context`.

File: packages/react/lib/common.js (hunk 3 above)
Comment (dmbch) [reply packages/react/lib/common.js]: Meanwhile, I have come to think that your initial implementation is the best we can do without changing the way `mixinable` works.

File: packages/react/README.md [hunk 4]
```diff
@@ -29,6 +29,25 @@ const App = () => <h1>Hello World!</h1>;
 export default render(<App />, createContext());
 ```
 
+## `render(reactElement, config)`
```
Comment (dmbch) [inline packages/react/README.md]: If we consider this the 'consumer' API, i.e. the most approachable way to use `render()`, why do we not mention it first? I really like this API and I think we ought to expose it most prominently.

File: packages/react/dom.js [hunk 5]
```diff
@@ -40,7 +42,7 @@ exports.contextDefinition = exports.ReactContext;
 
 exports.createContext = exports.combineContexts(exports.ReactContext);
 
-exports.render = function(reactElement, context) {
+exports.renderWithContext = function(reactElement, context) {
```
Comment (dmbch) [inline packages/react/dom.js]: Why do we export `renderWithContext`? Can we not just use `render`?

File: packages/react/README.md (hunk 4 above)
Comment (DJCordhose) [reply packages/react/README.md]: I had the same thoughts, but I had to somehow refer to the original render function for motivation, so it had to come first, I thought. Maybe not a valid point, not sure...

File: packages/react/dom.js (hunk 1 above)
Comment (DJCordhose) [reply packages/react/dom.js]: Yes, was broken, works fine with `isMixinable` :)

**Comment by DJCordhose** [comment]
As this is a non breaking change and commits are now squashed, I think it might be time for a merge if you want it. Details can be worked on once merged.

**Pull request merged**